io-amqp1_0-impl/src/main/java/org/apache/pulsar/ecosystem/io/amqp/AmqpSource.java

```diff
@@ -91,15 +92,17 @@ public void onMessage(Message message) {
                 if (config.isOnlyTextMessage()) {
                     record = new QpidJmsRecord(
                             Optional.empty(),
-                            ByteBuffer.wrap(((JmsTextMessage) message).getText().getBytes(StandardCharsets.UTF_8)));
+                            ByteBuffer.wrap(((JmsTextMessage) message).getText().getBytes(StandardCharsets.UTF_8)),
+                            message);
                 } else {
                     ByteBuf byteBuf = AmqpCodec.encodeMessage(
                             (AmqpJmsMessageFacade) ((JmsMessage) message).getFacade());
-                    record = new QpidJmsRecord(Optional.empty(), byteBuf.nioBuffer());
+                    record = new QpidJmsRecord(Optional.empty(), byteBuf.nioBuffer(), message);
                 }
                 pushSource.consume(record);
             } catch (Exception e) {
                 log.error("Failed to consume qpid jms message.", e);
+                throw new CouldNotProcessMessageException(e);
```

Comment:
Why do we throw this new exception?

Reply:
JmsMessageConsumer of the QpidJmsClient nacks a message when an exception is thrown. If we do not throw an exception the message will be consumed but not fully processed and be lost.

Reply:
Thanks for your explanation! SGTM.